File: src/output/router.ts
export interface CliWriteStream {
  write(chunk: string | Uint8Array): unknown;
}

export interface OutputRouterOptions {
  json: boolean;
  stdout?: CliWriteStream;
  stderr?: CliWriteStream;
}

export interface RoutedJsonResult {
  ok: boolean;
  status: string;
  command: string;
  message: string;
  dryRun: boolean;
}

export type JsonPayload = RoutedJsonResult | Record<string, unknown>;

type HumanTarget = "stdout" | "stderr";

const encoder = new TextEncoder();

function writeChunk(stream: CliWriteStream, message: string): void {
  stream.write(message);
}

function normalizeMessage(message: string): string {
  return message.endsWith("\n") ? message : `${message}\n`;
}

export function createOutputRouter({
  json,
  stdout = process.stdout,
  stderr = process.stderr
}: OutputRouterOptions) {
  return {
    isJson: json,
    writeHuman(message: string, target: HumanTarget = "stderr"): void {
      const stream = target === "stdout" ? stdout : stderr;
      writeChunk(stream, normalizeMessage(message));
    },
    writeJson(payload: JsonPayload): void {
      writeChunk(stdout, `${JSON.stringify(payload)}\n`);
    },
    writeStructured(
      payload: JsonPayload,
      message: string,
      target: HumanTarget = "stderr"
    ): void {
      if (json) {
        this.writeJson(payload);
        return;
      }

      this.writeHuman(message, target);
    },
    writeEncodedHuman(message: string, target: HumanTarget = "stderr"): void {
      const stream = target === "stdout" ? stdout : stderr;
      stream.write(encoder.encode(normalizeMessage(message)));
    }
  };
}

File: internal/output/router.go
package output

import (
	"encoding/json"
	"fmt"
	"io"
)

type HumanTarget string

const (
	StdoutTarget HumanTarget = "stdout"
	StderrTarget HumanTarget = "stderr"
)

type RoutedJSONResult struct {
	OK      bool   `json:"ok"`
	Status  string `json:"status"`
	Command string `json:"command"`
	Message string `json:"message"`
	DryRun  bool   `json:"dryRun"`
}

type Router struct {
	isJSON bool
	stdout io.Writer
	stderr io.Writer
}

func NewRouter(jsonMode bool, stdout, stderr io.Writer) *Router {
	return &Router{
		isJSON: jsonMode,
		stdout: stdout,
		stderr: stderr,
	}
}

func (r *Router) IsJSON() bool {
	return r.isJSON
}

func (r *Router) WriteHuman(message string, target HumanTarget) error {
	stream := r.stderr
	if target == StdoutTarget {
		stream = r.stdout
	}

	_, err := fmt.Fprintln(stream, message)
	return err
}

func (r *Router) WriteJSON(payload any) error {
	encoder := json.NewEncoder(r.stdout)
	return encoder.Encode(payload)
}

func (r *Router) WriteStructured(payload any, message string, target HumanTarget) error {
	if r.isJSON {
		return r.WriteJSON(payload)
	}

	return r.WriteHuman(message, target)
}
